### DISCUSSION TITLE
Error (crash) from plasticity // how to relax/bypass this error

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17743] [Q&A Modules: Solid mechanics] [answered]
Hi all, I tried to run a mining simulation with high initial stresses and rapid softening

but I often get this error message during the excavation step (reducing E and stress of a particular block)

> ++++++++++++++++++++++++
> Checking the derivatives
> ++++++++++++++++++++++++
> Debug Parameters are as follows
> stress = 
> 
> 
> *** ERROR ***
> The debug parameters have the wrong size

I tried to run it with the dynamic model/kernel with a lot of damping, it still crashes when I mine a large block within the geometry.

Is there any way that I can relax my simulation or bypass this error? (like relaxing the smoother/tolerance parameters)

```
[UserObjects]
  # [./mc_coh]
  #   type = TensorMechanicsHardeningConstant
  #   value = 5e6
  # [../]
  [./mc_coh]
    type = TensorMechanicsHardeningCubic
    value_0 = 5e6 
    value_residual = 5e5 
    internal_limit = 0.01 #reach at 1%
  [../]
  [./mc_phi]
    type = TensorMechanicsHardeningConstant
    value = 0.610865 #35deg
  [../]
  [./mc_psi]
    type = TensorMechanicsHardeningConstant
    value = 25
    convert_to_radians = true
  [../]
 ####### Mohr Coulomb
 [./mc]
   type = TensorMechanicsPlasticMohrCoulomb
   cohesion = mc_coh
   friction_angle = mc_phi
   dilation_angle = mc_psi
   yield_function_tolerance = 100
   internal_constraint_tolerance = 100
   mc_tip_smoother = 0.5e5
   mc_edge_smoother = 25
 [../]
[]
```

Thank you guys!

**Comment by Traiwit** [comment]
I also don't understand what does this error message mean? if I know it could probs work around it. Please help, thank you guys :)

**Comment by WilkAndy** [comment]
We need to know exactly when this occurs.  I think there are two possibilities:

1. During simulation.  You'll know this is the case if you also get the error `After reducing the stepsize to XXXX with original strain increment with L2norm YYYY  the returnMap algorithm failed` (where XXXX and YYYY depend on your problem).  The problem is that MOOSE cannot figure out the solution to the return-map process, which almost always means: (a) there is no solution because your plastic parameters, such as softening/hardening parameters, are not chosen wisely; or (b) your stresses are so extreme that you're getting precision loss, which can happen during the Newton procedure, and you might be able to choose different timestep sizes or damping parameters to avoid this; or (c) the stresses are in a configuration that i haven't coded for, which is my mistake.

2. During initialization.  You'll know this is the case when you don't get the aforementioned `After reducing...` error.  Choose `debug_fspb = none` to remove this error.

By the way, this error, particularly in case 2, could mean your initial stresses are inadmissible, which is potentially very bad, as some of the plasticity algorithms assume they are admissible, but your model might still produce reasonable results if you're lucky - you'll have to assess that for yourself.

Probably unrelated to this thread, but i wanted to mention it while i remember: I'm pretty sure you're changing the Young's modulus to simulate tunnel excavation.  Make sure you have one initial timestep with it set to its unmodified value, otherwise you'll be getting the wrong answer.  I can explain why in another discussion thread, but, if your `elasticity_tensor_prefactor = if(t>0,1E-3,1)` just make sure you have `start_time=-1` `dt=1` or something like that.

**Reply by Traiwit** [reply]
Hi @WilkAndy it crashes in the middle of simulation, sometimes the excavation of the 2nd block, or 3rd stope, etc depending on the plasticity criteria I have.

![Screenshot from 2021-05-05 09-52-12](https://user-images.githubusercontent.com/79562439/117085281-ad553000-ad8c-11eb-9fcb-b7df1b2aa80f.png)

You have given out quite a lot of information here, I will try to work on them and report to you back soon.

note: on the last point, I always let the system stabilise for 1 sec before start excavation, that shouldn't be a problem.

**Reply by Traiwit** [reply]
Hi @WilkAndy, one thing I observed, the simulation crashes when the plastic strain (mc_int/plastic_internal_parameter) is getting closer to the softening residual strain, in this case 1%.

  [./mc_coh]
    type = TensorMechanicsHardeningCubic
    value_0 = 5e6 
    value_residual = 5e5 
    internal_limit = 0.01 #reach at 1%
  [../]

I tried to turn off the initial stress, so the simulation run perfectly fine, however at the last time step I see the max (mc_int/plastic_internal_parameter) is 0.02% which is far away from 1%.

**Reply by Traiwit** [reply]
I also see in mohr_coulomb/small_deform_hard_cubic.i under [Materials/ComputeMultiPlasticityStress] it contains quite a number of debug parameters, will these be helpful for my simulation?

```
  [./mc]
    type = ComputeMultiPlasticityStress
    block = 0
    ep_plastic_tolerance = 1E-9
    plastic_models = mc
    debug_fspb = crash
    debug_jac_at_stress = '10 1 2 1 10 3 2 3 10'
    debug_jac_at_pm = 1
    debug_jac_at_intnl = 1E-4
    debug_stress_change = 1E-5
    debug_pm_change = 1E-6
    debug_intnl_change = 1E-8
  [../]
```

**Reply by WilkAndy** [reply]
I think these won't be helpful for you.  They're more for developers to check that derivatives and so on are actually correct.

**Reply by Traiwit** [reply]
Thanks @WilkAndy! I really have no idea what's going on here, I probs need to work on a much simpler geometry to find out what's the actual problem.

**Comment by WilkAndy** [comment, marked as answer]
Three other things you're probably going to experiment with:

1. Try setting the dilation angle equal to the friction angle.  Eventually you'll want to set them different, but for your debugging dil=fric is a much easier problem to solve for MOOSE.
2. Choose different `yield_function_tolerance`, `internal_constraint_tolerance` and `ep_plastic_tolerance`.  `yield_function_tolerance` should usually be set lower than any meaningful stresses.  Eg, if your stresses are measured in Pascals, then `yield_function_tolerance = 1E-7` is fine, since it means "1E-7 Pascals is a small number".  Do not set it too low, because MOOSE may run out of precision, but do not set it too high (like your value of 100) because it allows MOOSE to be too lazy and your stresses will end up being crazy at some stage.  I usually start at 1E-14 * youngs_modulus, and then move downwards to see if MOOSE can still solve the problem.  `ep_plastic_tolerance` should be set at around `yield_function_tolerance` divided by youngs_modulus, but no lower than 1E-12, otherwise you'll loose precision.  `internal_constraint_tolerance` is usually less important, but should certainly be set less than your current value of 100.  Currently you're saying "measure plastic strain with an accuracy of 100", which is obviously not optimal, since your hardening occurs over the region to 0.01.  So use something like 1E-12 for this parameter (but don't panic - this parameter doesn't usually make much difference to the final results).
3. Try using DruckerPrager plasticity.  It's numerically easier than MC.

**Comment by PEI0214** [comment]
Hello, I also met a similar situation, how did you solve it?
